{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "dcae453d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Data reuse status</th>\n",
       "      <th>Publication link</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dataset name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Brca</th>\n",
       "      <td>breast cancer</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/33212010/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ccrcc</th>\n",
       "      <td>clear cell renal cell carcinoma (kidney)</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/31675502/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Colon</th>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/31031003/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Endometrial</th>\n",
       "      <td>endometrial carcinoma (uterine)</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/32059776/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gbm</th>\n",
       "      <td>glioblastoma</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/33577785/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hnscc</th>\n",
       "      <td>head and neck squamous cell carcinoma</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/33417831/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lscc</th>\n",
       "      <td>lung squamous cell carcinoma</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/34358469/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Luad</th>\n",
       "      <td>lung adenocarcinoma</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/32649874/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ovarian</th>\n",
       "      <td>high grade serous ovarian cancer</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/27372738/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pdac</th>\n",
       "      <td>pancreatic ductal adenocarcinoma</td>\n",
       "      <td>no restrictions</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/34534465/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UcecConf</th>\n",
       "      <td>endometrial confirmatory carcinoma</td>\n",
       "      <td>password access only</td>\n",
       "      <td>unpublished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GbmConf</th>\n",
       "      <td>glioblastoma confirmatory</td>\n",
       "      <td>password access only</td>\n",
       "      <td>unpublished</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Description     Data reuse status  \\\n",
       "Dataset name                                                                   \n",
       "Brca                                     breast cancer       no restrictions   \n",
       "Ccrcc         clear cell renal cell carcinoma (kidney)       no restrictions   \n",
       "Colon                                colorectal cancer       no restrictions   \n",
       "Endometrial            endometrial carcinoma (uterine)       no restrictions   \n",
       "Gbm                                       glioblastoma       no restrictions   \n",
       "Hnscc            head and neck squamous cell carcinoma       no restrictions   \n",
       "Lscc                      lung squamous cell carcinoma       no restrictions   \n",
       "Luad                               lung adenocarcinoma       no restrictions   \n",
       "Ovarian               high grade serous ovarian cancer       no restrictions   \n",
       "Pdac                  pancreatic ductal adenocarcinoma       no restrictions   \n",
       "UcecConf            endometrial confirmatory carcinoma  password access only   \n",
       "GbmConf                      glioblastoma confirmatory  password access only   \n",
       "\n",
       "                                       Publication link  \n",
       "Dataset name                                             \n",
       "Brca          https://pubmed.ncbi.nlm.nih.gov/33212010/  \n",
       "Ccrcc         https://pubmed.ncbi.nlm.nih.gov/31675502/  \n",
       "Colon         https://pubmed.ncbi.nlm.nih.gov/31031003/  \n",
       "Endometrial   https://pubmed.ncbi.nlm.nih.gov/32059776/  \n",
       "Gbm           https://pubmed.ncbi.nlm.nih.gov/33577785/  \n",
       "Hnscc         https://pubmed.ncbi.nlm.nih.gov/33417831/  \n",
       "Lscc          https://pubmed.ncbi.nlm.nih.gov/34358469/  \n",
       "Luad          https://pubmed.ncbi.nlm.nih.gov/32649874/  \n",
       "Ovarian       https://pubmed.ncbi.nlm.nih.gov/27372738/  \n",
       "Pdac          https://pubmed.ncbi.nlm.nih.gov/34534465/  \n",
       "UcecConf                                    unpublished  \n",
       "GbmConf                                     unpublished  "
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import cptac\n",
    "\n",
    "cptac.list_datasets()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "b5db7a96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                          \r"
     ]
    }
   ],
   "source": [
    "cptac.download(dataset='Ccrcc')\n",
    "\n",
    "ccrcc = cptac.Ccrcc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "869b90ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "f2b029f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Find the top 5 most differentially expressed proteins between Stage I and Stage III patients\n",
    "\n",
    "clinical_data = ccrcc.get_clinical()\n",
    "rna_data = ccrcc.get_transcriptomics()\n",
    "protein_data = ccrcc.get_proteomics()\n",
    "\n",
    "protein_data.columns = protein_data.columns.get_level_values(0) \n",
    "protein_data = protein_data.dropna(axis=1)\n",
    "\n",
    "tumor_mask = clinical_data.loc[:, \"Sample_Tumor_Normal\"] == \"Tumor\"\n",
    "clinical_data = clinical_data.loc[tumor_mask, :]\n",
    "\n",
    "stage_I_mask = clinical_data.loc[:, \"tumor_stage_pathological\"] == \"Stage I\"\n",
    "stage_III_mask = clinical_data.loc[:, \"tumor_stage_pathological\"] == \"Stage III\"\n",
    "\n",
    "stage_I_patients = (clinical_data.index[stage_I_mask]).dropna()\n",
    "stage_III_patients = (clinical_data.index[stage_III_mask]).dropna()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "ce2ecb95",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = [stage_I_patients.values, stage_III_patients.values]\n",
    "\n",
    "protein_stage_I = protein_data.loc[stage_I_patients, :]\n",
    "protein_stage_III = protein_data.loc[stage_III_patients, :]\n",
    "\n",
    "protein_diff = protein_stage_I.mean() - protein_stage_III.mean()\n",
    "protein_diff = protein_diff.abs()\n",
    "protein_diff = protein_diff.sort_values()\n",
    "protein_data = protein_data.loc[tumor_mask, :]\n",
    "\n",
    "protein_list = [\"FTL\", \"HBZ\", \"HBA2\", \"CMA1\", \"HBB\"]\n",
    "\n",
    "top_5_protein = protein_data.loc[:, protein_list]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "6cf92a23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Name\n",
       "ACADM      0.000011\n",
       "SMCHD1     0.000045\n",
       "ZC3H11A    0.000054\n",
       "EXOSC3     0.000071\n",
       "NAXE       0.000083\n",
       "             ...   \n",
       "HBB        0.557303\n",
       "CMA1       0.583464\n",
       "HBA2       0.589100\n",
       "HBZ        0.603453\n",
       "FTL        0.862165\n",
       "Length: 6665, dtype: float64"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_diff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "00683140",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Windows\\anaconda3\\envs\\qbio_490_nathanyoon\\lib\\site-packages\\pandas\\core\\internals\\blocks.py:366: RuntimeWarning: invalid value encountered in log2\n",
      "  result = func(self.values, **kwargs)\n",
      "C:\\Users\\Windows\\AppData\\Local\\Temp\\ipykernel_16036\\3226768701.py:9: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  rna_data = rna_data[tumor_mask]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Name\n",
       "SNRPN     0.000003\n",
       "RBM14     0.000015\n",
       "PMPCB     0.000017\n",
       "VPS35     0.000020\n",
       "SNRPD3    0.000021\n",
       "            ...   \n",
       "PTTG1     0.410133\n",
       "CBWD7     0.410572\n",
       "TNC       0.422949\n",
       "CCRL2     0.502770\n",
       "CC2D2B    0.510390\n",
       "Length: 9806, dtype: float64"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_data = np.log2(rna_data).dropna(axis=1)\n",
    "rna_stage_I = rna_data.loc[stage_I_patients, :]\n",
    "rna_stage_III = rna_data.loc[stage_III_patients, :]\n",
    "\n",
    "rna_diff = rna_diff.dropna()\n",
    "rna_diff = rna_stage_I.mean() - rna_stage_III.mean()\n",
    "rna_diff = rna_diff.abs()\n",
    "rna_diff = rna_diff.sort_values()\n",
    "rna_data = rna_data[tumor_mask]\n",
    "\n",
    "rna_list = [\"CC2D2B\", \"CCRL2\", \"TNC\", \"CBWD7\", \"PTTG1\"]\n",
    "\n",
    "top_5_rna = rna_data.loc[:, rna_list]\n",
    "rna_diff\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "e70f8812",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Create a new dataframe: rows = patients, cols = expression values of genes\n",
    "top_10 = pd.concat([top_5_protein, top_5_rna], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "a0754b8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>FTL</th>\n",
       "      <th>HBZ</th>\n",
       "      <th>HBA2</th>\n",
       "      <th>CMA1</th>\n",
       "      <th>HBB</th>\n",
       "      <th>CC2D2B</th>\n",
       "      <th>CCRL2</th>\n",
       "      <th>TNC</th>\n",
       "      <th>CBWD7</th>\n",
       "      <th>PTTG1</th>\n",
       "      <th>tumor_stage_pathological</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00004</th>\n",
       "      <td>-1.724339</td>\n",
       "      <td>-0.370098</td>\n",
       "      <td>-0.786307</td>\n",
       "      <td>-0.918273</td>\n",
       "      <td>-0.762287</td>\n",
       "      <td>0.175725</td>\n",
       "      <td>0.893170</td>\n",
       "      <td>1.226618</td>\n",
       "      <td>0.410942</td>\n",
       "      <td>1.080103</td>\n",
       "      <td>Stage III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00010</th>\n",
       "      <td>-0.363228</td>\n",
       "      <td>0.240576</td>\n",
       "      <td>-0.081263</td>\n",
       "      <td>0.747860</td>\n",
       "      <td>-0.034944</td>\n",
       "      <td>0.027322</td>\n",
       "      <td>-0.180820</td>\n",
       "      <td>1.006059</td>\n",
       "      <td>0.288196</td>\n",
       "      <td>0.702747</td>\n",
       "      <td>Stage I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00011</th>\n",
       "      <td>-0.977364</td>\n",
       "      <td>-0.087641</td>\n",
       "      <td>-0.418663</td>\n",
       "      <td>-0.255054</td>\n",
       "      <td>-0.371128</td>\n",
       "      <td>0.280647</td>\n",
       "      <td>0.151815</td>\n",
       "      <td>0.561198</td>\n",
       "      <td>0.508859</td>\n",
       "      <td>2.292182</td>\n",
       "      <td>Stage IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00026</th>\n",
       "      <td>1.301930</td>\n",
       "      <td>0.981930</td>\n",
       "      <td>0.780756</td>\n",
       "      <td>0.049086</td>\n",
       "      <td>0.855097</td>\n",
       "      <td>0.929820</td>\n",
       "      <td>0.144756</td>\n",
       "      <td>1.776796</td>\n",
       "      <td>0.672007</td>\n",
       "      <td>1.047153</td>\n",
       "      <td>Stage I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00079</th>\n",
       "      <td>-1.496648</td>\n",
       "      <td>-0.441854</td>\n",
       "      <td>-1.014088</td>\n",
       "      <td>-0.620829</td>\n",
       "      <td>-0.939448</td>\n",
       "      <td>0.209431</td>\n",
       "      <td>0.161999</td>\n",
       "      <td>1.343163</td>\n",
       "      <td>-0.542388</td>\n",
       "      <td>2.404735</td>\n",
       "      <td>Stage III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01646</th>\n",
       "      <td>-1.817120</td>\n",
       "      <td>-0.724017</td>\n",
       "      <td>-1.180566</td>\n",
       "      <td>-0.000699</td>\n",
       "      <td>-1.127855</td>\n",
       "      <td>0.333554</td>\n",
       "      <td>-0.175023</td>\n",
       "      <td>2.115398</td>\n",
       "      <td>-2.178603</td>\n",
       "      <td>0.759980</td>\n",
       "      <td>Stage III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01648</th>\n",
       "      <td>0.459593</td>\n",
       "      <td>-0.396638</td>\n",
       "      <td>-0.672383</td>\n",
       "      <td>-0.611409</td>\n",
       "      <td>-0.605164</td>\n",
       "      <td>0.346245</td>\n",
       "      <td>2.099780</td>\n",
       "      <td>1.693689</td>\n",
       "      <td>-0.309480</td>\n",
       "      <td>1.305223</td>\n",
       "      <td>Stage II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01649</th>\n",
       "      <td>-1.066512</td>\n",
       "      <td>-0.109630</td>\n",
       "      <td>-0.584516</td>\n",
       "      <td>0.322161</td>\n",
       "      <td>-0.527777</td>\n",
       "      <td>0.990790</td>\n",
       "      <td>0.241445</td>\n",
       "      <td>2.320249</td>\n",
       "      <td>-1.896508</td>\n",
       "      <td>0.567970</td>\n",
       "      <td>Stage III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01651</th>\n",
       "      <td>-1.895398</td>\n",
       "      <td>-0.096922</td>\n",
       "      <td>-0.649448</td>\n",
       "      <td>-0.302394</td>\n",
       "      <td>-0.571897</td>\n",
       "      <td>-0.155102</td>\n",
       "      <td>-0.035762</td>\n",
       "      <td>1.822354</td>\n",
       "      <td>0.081185</td>\n",
       "      <td>1.608496</td>\n",
       "      <td>Stage II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01808</th>\n",
       "      <td>1.895297</td>\n",
       "      <td>2.056017</td>\n",
       "      <td>1.485439</td>\n",
       "      <td>-1.125322</td>\n",
       "      <td>1.538612</td>\n",
       "      <td>0.701076</td>\n",
       "      <td>0.332174</td>\n",
       "      <td>1.806733</td>\n",
       "      <td>0.796962</td>\n",
       "      <td>0.959540</td>\n",
       "      <td>Stage I</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name             FTL       HBZ      HBA2      CMA1       HBB    CC2D2B  \\\n",
       "Patient_ID                                                               \n",
       "C3L-00004  -1.724339 -0.370098 -0.786307 -0.918273 -0.762287  0.175725   \n",
       "C3L-00010  -0.363228  0.240576 -0.081263  0.747860 -0.034944  0.027322   \n",
       "C3L-00011  -0.977364 -0.087641 -0.418663 -0.255054 -0.371128  0.280647   \n",
       "C3L-00026   1.301930  0.981930  0.780756  0.049086  0.855097  0.929820   \n",
       "C3L-00079  -1.496648 -0.441854 -1.014088 -0.620829 -0.939448  0.209431   \n",
       "...              ...       ...       ...       ...       ...       ...   \n",
       "C3N-01646  -1.817120 -0.724017 -1.180566 -0.000699 -1.127855  0.333554   \n",
       "C3N-01648   0.459593 -0.396638 -0.672383 -0.611409 -0.605164  0.346245   \n",
       "C3N-01649  -1.066512 -0.109630 -0.584516  0.322161 -0.527777  0.990790   \n",
       "C3N-01651  -1.895398 -0.096922 -0.649448 -0.302394 -0.571897 -0.155102   \n",
       "C3N-01808   1.895297  2.056017  1.485439 -1.125322  1.538612  0.701076   \n",
       "\n",
       "Name           CCRL2       TNC     CBWD7     PTTG1 tumor_stage_pathological  \n",
       "Patient_ID                                                                   \n",
       "C3L-00004   0.893170  1.226618  0.410942  1.080103                Stage III  \n",
       "C3L-00010  -0.180820  1.006059  0.288196  0.702747                  Stage I  \n",
       "C3L-00011   0.151815  0.561198  0.508859  2.292182                 Stage IV  \n",
       "C3L-00026   0.144756  1.776796  0.672007  1.047153                  Stage I  \n",
       "C3L-00079   0.161999  1.343163 -0.542388  2.404735                Stage III  \n",
       "...              ...       ...       ...       ...                      ...  \n",
       "C3N-01646  -0.175023  2.115398 -2.178603  0.759980                Stage III  \n",
       "C3N-01648   2.099780  1.693689 -0.309480  1.305223                 Stage II  \n",
       "C3N-01649   0.241445  2.320249 -1.896508  0.567970                Stage III  \n",
       "C3N-01651  -0.035762  1.822354  0.081185  1.608496                 Stage II  \n",
       "C3N-01808   0.332174  1.806733  0.796962  0.959540                  Stage I  \n",
       "\n",
       "[110 rows x 11 columns]"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. Create a separate list of the patients' cancer stages\n",
    "\n",
    "top_10[\"tumor_stage_pathological\"] = clinical_data.loc[:, \"tumor_stage_pathological\"]\n",
    "\n",
    "top_10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "3b8d4614",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4. Scale and encode data\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from umap import UMAP\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "encoder = LabelEncoder()\n",
    "unencoded_columns = top_10.loc[:, \"tumor_stage_pathological\"]\n",
    "encoded_columns = encoder.fit_transform(unencoded_columns)\n",
    "\n",
    "scaler = StandardScaler()\n",
    "\n",
    "scaled_data = scaler.fit_transform(top_10.iloc[:, 0:10])\n",
    "embedding_scaled = UMAP().fit_transform(scaled_data)\n",
    "\n",
    "data = scaled_data[:, 0:9]\n",
    "target = scaled_data[:, 10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "5ebff6ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(110, 11)"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_10.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "c8894557",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5. Create a train test split\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "target = top_10.loc[:, \"tumor_stage_pathological\"]\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(scaled_data, target, train_size = 0.7)\n",
    "\n",
    "scaler = StandardScaler()\n",
    "\n",
    "scaler.fit(X_train)\n",
    "\n",
    "X_train = scaler.transform(X_train)\n",
    "\n",
    "X_test = scaler.transform(X_test)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "52dedede",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6. Test all 4 classification models\n",
    "\n",
    "from sklearn.neighbors import KNeighborsClassifier # default number of neighbors looked at is 5\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.naive_bayes import GaussianNB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "03f8b7c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using the KNeighborsClassifier, the model was able to accurately predict 63.63636363636363% of the test set.\n",
      "Using the DecisionTreeClassifier, the model was able to accurately predict 36.36363636363637% of the test set.\n",
      "Using the MLPClassifier, the model was able to accurately predict 54.54545454545454% of the test set.\n",
      "Using the GaussianNB, the model was able to accurately predict 54.54545454545454% of the test set.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Windows\\anaconda3\\envs\\qbio_490_nathanyoon\\lib\\site-packages\\sklearn\\neural_network\\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "classifier = KNeighborsClassifier()\n",
    "classifier.fit(X_train, y_train)\n",
    "y_pred = classifier.predict(X_test)\n",
    "\n",
    "accuracy = sum(y_pred == y_test) / len(y_test)\n",
    "\n",
    "print(f'Using the KNeighborsClassifier, the model was able to accurately predict {accuracy * 100}% of the test set.')\n",
    "\n",
    "classifier = DecisionTreeClassifier()\n",
    "classifier.fit(X_train, y_train)\n",
    "y_pred = classifier.predict(X_test)\n",
    "\n",
    "accuracy = sum(y_pred == y_test) / len(y_test)\n",
    "\n",
    "print(f'Using the DecisionTreeClassifier, the model was able to accurately predict {accuracy * 100}% of the test set.')\n",
    "\n",
    "classifier = MLPClassifier()\n",
    "classifier.fit(X_train, y_train)\n",
    "y_pred = classifier.predict(X_test)\n",
    "\n",
    "accuracy = sum(y_pred == y_test) / len(y_test)\n",
    "\n",
    "print(f'Using the MLPClassifier, the model was able to accurately predict {accuracy * 100}% of the test set.')\n",
    "\n",
    "classifier = GaussianNB()\n",
    "classifier.fit(X_train, y_train)\n",
    "y_pred = classifier.predict(X_test)\n",
    "\n",
    "accuracy = sum(y_pred == y_test) / len(y_test)\n",
    "\n",
    "print(f'Using the GaussianNB, the model was able to accurately predict {accuracy * 100}% of the test set.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "1ca07a8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7. Compare 4 mean accuracies.\n",
    "\n",
    "# According to the results,\n",
    "# KNeighbors    - 63.6% - \"Best\"\n",
    "# DecisionTree  - 36.3% - \"Worst\"\n",
    "# MLPClassifier - 54.5%\n",
    "# GaussianNB    - 54.5%\n",
    "\n",
    "# If I did my code right, none of these models seem to predict the data very well.\n",
    "\n",
    "# All of these values imply the set size is a multiple of 11 - while this initially caused confusion if the final \n",
    "# tumor_stage_pathological was included (making number of columns = 11), I also realized the number of rows (X values)\n",
    "# is 110, so maybe there is no problem?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eb5aba9",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
